Code/Tools/AssetProcessor/Platform/Linux/native/FileWatcher/FileWatcher_linux.cpp

```diff
@@ -154,90 +289,178 @@ bool FileWatcher::PlatformStart()
+                break;
+            }
+            
+            if (!bytesRead)
+            {
+                continue;
+            }
         }
+
+        if (m_shutdownThreadSignal)
+        {
+            break;
+        }
+
+        cycleCount++;
+
         for (size_t index=0; index<bytesRead;)
         {
             const auto* event = reinterpret_cast<inotify_event*>(&eventBuffer[index]);
 
-            if (event->mask & (IN_CREATE | IN_DELETE | IN_MODIFY | IN_MOVE))
+            if (event->mask & (IN_CREATE | IN_DELETE | IN_MODIFY | IN_MOVE | IN_DELETE_SELF | IN_MOVE_SELF ))
             {
-                const QString pathStr = QDir(m_platformImpl->m_handleToFolderMap[event->wd]).absoluteFilePath(event->name);
+                // note that the event->name coming in is relative to the thing being watched.  Since we watch folders,
+                // for the folder itself, this will be blank, for files in it, it will be the file name.
+                QDir watchedDir(m_platformImpl->m_handleToFolderMap[event->wd]);
+                const QString pathStr = watchedDir.absoluteFilePath(event->name);
 
                 if (event->mask & (IN_CREATE | IN_MOVED_TO))
                 {
-                    if (event->mask & IN_ISDIR)
-                    {
-                        // New Directory, see if it should be added to the watched directories
-                        // It is only added if it is a child of a recursively watched directory
-                        const auto found = AZStd::find_if(begin(m_folderWatchRoots), end(m_folderWatchRoots), [this, event](const WatchRoot& watchRoot)
+                    DEBUG_FILEWATCHER("notify event is IN_CREATE | IN_MOVED_TO (flags 0x%08x) %s (from '%s') cycle: %i \n", event->mask, pathStr.toUtf8().constData(), event->name, cycleCount);
+                    const auto found = AZStd::find_if(begin(m_folderWatchRoots), end(m_folderWatchRoots), [this, event](const WatchRoot& watchRoot)
                         {
                             return watchRoot.m_directory == m_platformImpl->m_handleToFolderMap[event->wd];
                         });
 
-                        // If the path is not in m_folderWatchRoots, it must
-                        // be a new subdirectory of a subdirectory of some
-                        // other root that is being watched recursively.
-                        // Maintain the recursive nature of that root.
-                        const bool shouldAddFolder = (found == end(m_folderWatchRoots)) ? true : found->m_recursive;
-
-                        if (shouldAddFolder)
+                    bool isChildOfRootFolder = found != end(m_folderWatchRoots);
+                    bool isChildOfRecursiveRootFolder = isChildOfRootFolder && found->m_recursive;
+                    
+                    if (event->mask & IN_ISDIR)
+                    {
+                        // for directories, we only only care about create or delete, not modify
```

Comment:
nitpick: typo double "only"
```suggestion
                        // for directories, we only care about create or delete, not modify
```